test/index.spec.js

```diff
@@ -236,3 +247,114 @@ it('should fulfill the promise the array of expected actions is contained in the
 
     return promise;
 });
+
+it('should teardown correctly when promise fulfills', async () => {
+    const mockStore = createMockStore([thunkMiddleware]);
+    const originalSubscribe = mockStore.subscribe;
+    let timeoutId;
+    let unsubscribeMock;
+
+    mockStore.subscribe = (callback) => {
+        const unsubscribe = originalSubscribe.call(mockStore, callback);
+
+        unsubscribeMock = jest.fn(() => unsubscribe());
+
+        return unsubscribeMock;
+    };
+    global.setTimeout = (callback, ms) => {
+        timeoutId = originalSetTimeout.call(global, callback, ms);
+
+        return timeoutId;
+    };
+    global.clearTimeout = jest.fn((timer) => originalClearTimeout.call(global, timer));
+
+    mockStore.dispatch(asyncActionsCreator([action1]));
+
+    await waitForActions(mockStore, action1);
+
+    expect(clearTimeout).toHaveBeenCalledWith(timeoutId);
+    expect(unsubscribeMock).toHaveBeenCalled();
+});
+
+it('should teardown correctly when promise rejects via timeout', async () => {
+    expect.assertions(1);
+    const mockStore = createMockStore([thunkMiddleware]);
+    const originalSubscribe = mockStore.subscribe;
+    let unsubscribeMock;
+
+    mockStore.subscribe = (callback) => {
+        const unsubscribe = originalSubscribe.call(mockStore, callback);
+
+        unsubscribeMock = jest.fn(() => unsubscribe());
+
+        return unsubscribeMock;
+    };
+
+    mockStore.dispatch(asyncActionsCreator([action1]));
+
+    try {
+        await waitForActions(mockStore, [action1, action2]);
+    } catch (err) {
+        expect(unsubscribeMock).toHaveBeenCalled();
+    }
+});
+
+it('should teardown correctly when promise rejects via cancelation', () => {
+    expect.assertions(1);
+    const mockStore = createMockStore([thunkMiddleware]);
+    const originalSubscribe = mockStore.subscribe;
+    let timeoutId;
+    let unsubscribeMock;
+
+    mockStore.subscribe = (callback) => {
+        const unsubscribe = originalSubscribe.call(mockStore, callback);
+
+        unsubscribeMock = jest.fn(() => unsubscribe());
+
+        return unsubscribeMock;
+    };
+    global.setTimeout = (callback, ms) => {
```

Comment:
I think we could have used `jest.spyOn(global, 'clearTimeout')` and `expect(clearTimeout).toHaveBeenCalledTimes(1)` here and maybe in other cases? There's no need to actually check the timer id, at least I think this is enough.

Reply:
Agreed. 😄